modules/java/CMakeLists.txt

```diff
@@ -84,42 +66,40 @@ endif()
 
   include(UseJava)
 
+if ("${Java_VERSION}" VERSION_LESS 1.8)
   add_jar(syslog-ng-core
       SOURCES ${JAVA_SOURCES}
       OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR}
   )
 
-  gen_java_header(
-      ${CMAKE_CURRENT_BINARY_DIR}/org_syslog_ng_LogMessage.h
-      ${CMAKE_CURRENT_BINARY_DIR}/syslog-ng-core.jar
-      ${CMAKE_CURRENT_BINARY_DIR}
-      "org.syslog_ng.LogMessage"
-      "syslog-ng-core"
+  #Both exists in different version of cmake
+  get_property(SYSLOG_NG_CLASSPATH TARGET syslog-ng-core PROPERTY CLASS_DIR)
+  if("${SYSLOG_NG_CLASSPATH}" STREQUAL "")
+    get_property(SYSLOG_NG_CLASSPATH TARGET syslog-ng-core PROPERTY CLASSDIR)
+  endif()
+
+  create_javah(GENERATED_FILES BUILT_HEADERS
+        CLASSES    org.syslog_ng.LogMessage org.syslog_ng.InternalMessageSender org.syslog_ng.LogDestination org.syslog_ng.LogTemplate org.syslog_ng.LogPipe
+        CLASSPATH ${SYSLOG_NG_CLASSPATH}
+        OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR}
+        DEPENDS syslog-ng-core
   )
 
-  gen_java_header(
-      ${CMAKE_CURRENT_BINARY_DIR}/org_syslog_ng_InternalMessageSender.h
-      ${CMAKE_CURRENT_BINARY_DIR}/syslog-ng-core.jar
-      ${CMAKE_CURRENT_BINARY_DIR}
-      "org.syslog_ng.InternalMessageSender"
-      "syslog-ng-core"
-  )
+  set_source_files_properties(${BUILT_HEADERS} PROPERTIES GENERATED TRUE)
+else()
+  #The GENERATE_NATIVE_HEADERS variable introduced only in cmake 3.11.1
+  #This is just a dirty workaround for that:
+  set(CMAKE_JAVA_COMPILE_FLAGS "${CMAKE_JAVA_COMPILE_FLAGS};-h;${CMAKE_CURRENT_BINARY_DIR};")
 
-  gen_java_header(
-      ${CMAKE_CURRENT_BINARY_DIR}/org_syslog_ng_LogDestination.h
-      ${CMAKE_CURRENT_BINARY_DIR}/syslog-ng-core.jar
-      ${CMAKE_CURRENT_BINARY_DIR}
-      "org.syslog_ng.LogDestination"
-      "syslog-ng-core"
+  add_jar(syslog-ng-core
+      SOURCES ${JAVA_SOURCES}
+      OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR}
+      GENERATE_NATIVE_HEADERS syslog-ng-core-headers
   )
 
-  gen_java_header(
-      ${CMAKE_CURRENT_BINARY_DIR}/org_syslog_ng_LogTemplate.h
-      ${CMAKE_CURRENT_BINARY_DIR}/syslog-ng-core.jar
-      ${CMAKE_CURRENT_BINARY_DIR}
-      "org.syslog_ng.LogTemplate"
-      "syslog-ng-core"
-  )
+  set(BUILT_HEADERS "${CMAKE_CURRENT_BINARY_DIR}/syslog-ng-core.jar")
```

Comment:
It looks strange to call the jar as BUILT_HEADERS, which originally contained the headers. Can you just explicitely call set_source_files_properties on the jar?

Reply:
Hmm, now I kinda do not remember why was it written in this way. But I do not think it is right.

Reply:
@Kokan Could you have a look at this again, so we can finish the PR?

Reply:
@MrAnno @furiel sorry for the delay